### PR TITLE
Adjust the order of evaluation for implicit Index/Range support to follow the recent LDM decision.

### LINKED ISSUE #57349: Inconsistent order of evaluation between indexing with System.Index and System.Range
**Version Used**: Roslyn main (20 Oct 2021)

**Steps to Reproduce**:

Run the following code:
```c#
using System;
using System.Collections;
using System.Collections.Generic;

class Program
{
    static CustomList list;

    public static void Main()
    {
        list = new CustomList { "first", "second", "last" };
        Console.WriteLine(list[AllButLast()]);
        list = new CustomList { "first", "second", "last" };
        Console.WriteLine(list[Last()]); // outputs "last" with VS2019 and "NewLast" with VS2022
    }

    static Range AllButLast()
    {
        list.Add("New");
        list.Add("NewLast");
        return 0..^1;
    }

    static Index Last()
    {
        list.Add("New");
        list.Add("NewLast");
        return ^1;
    }
}

internal class CustomList : IEnumerable<string>
{
    List<string> l = new List<string>();

    public void Add(string arg) => l.Add(arg);

    public int Count => l.Count;
    public string this[int index] => l[index];

    public CustomList Slice(int start, int length)
    {
        CustomList result = new CustomList();
        for (int i = 0; i < length; i++)
        {
            result.Add(this[start + i]);
        }
        return result;
    }

    public IEnumerator<string> GetEnumerator()
    {
        return l.GetEnumerator();
    }

    IEnumerator IEnumerable.GetEnumerator()
    {
        return l.GetEnumerator();
    }

    public override string ToString()
    {
        return string.Join(", ", l);
    }
}
```

**Expected Behavior**:
The methods `AllButLast`/`Last` should be called before the indexing operator retrieves `list.Count`.
`Console.WriteLine(list[AllButLast()])` should output `first, second, last, New`.
`Console.WriteLine(list[Last()]);` should output `NewLast`.

**Actual Behavior**:
VS2019 retrieves `list.Count` before calling `AllButLast`/`Last`, output:
```
first, second
last
```

Roslyn main (20 Oct 2021) retrieves `list.Count` before calling `AllButLast`, but after calling `Last`, output:
```
first, second
NewLast
```

[See on SharpLab](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK7wCYgNQB8ACATAIwCwAUPgAwAE+xAdAMID2ANmwKYDGCULMAM4BuCtTrEALKPJjCEgOwUA3hRrqJANhpNUghCzAAZKPpptTCGWo1FCN9avIaX5yzQC8NGJwDuOvQNjd2UaACIAMyhkfTC4cMEeAQw48LYIWJoAXxlXW2IATgAKC30AbQBBDgAhVAQjDIQigEoAXWbcvLczLx9/XX1DEzNQyOjY+LDE7mTUsPTMnIdXemLShDKG/Rb24RoAen2aFjqABzrBNMawml8oBAALGgA1AGVCKkKaCBgMcIA5PxbBA3O6PF7vKiEezODRZCjLejaABKPwA5pwaFU2LV6o0WssnF1uggGBUMBgimFAb4wh1li51mSKVSacC6Z08vgFDQqAwGAA9Yic7II2HqJE0ACSv04AA8aMCCeKaESukzyZTqX4OQyNBqWdrfOz6SqXNyaEKRfDyDaKLAEJxkDAIGw6PIBkFhggaCAJABmAA89CoAD4VMtvcHiGHzJ5vH5FZZo2GWtYVfh/XRJFiWSHvsg0c1PKHzMzKRBC6bEVmHToTjAfR5S2xmA2rDWJLRHqYynXYBh5a0S+Y+7K5a10+as56hu5XhZuJwinX9JWEPE61wYGjHs1CXr1LPgmZkJxBKg2E2E/1AnPtqbiREWMgaCvGzQoPGqHsv4HzJwO6PL+2DYPuZoaGqxLqGeF5XuWRQ9oIZRrsgPrYJ+uyHqKEESjysGXh2Kp2hmWZSgAojAqBgE6EAGMgKalgA4pwCCUdRtH0cqLhQSsPKtixbFUTRyB0S+abLCRLgUcJnEvgGDCCexIlicg3GQdhFoCaxylyWpj5wmK07HAAbk6yBQIOXY0AAKiwrwIBZO7qY4mk8iGDAAFIsLAVKTPEbAGeoNo2kAA==)

### PR DESCRIPTION
**Pull request by AlekseyTs**
LDM notes - https://github.com/dotnet/csharplang/blob/main/meetings/2021/LDM-2021-11-01.md.
Proposed specification change - https://github.com/dotnet/csharplang/pull/5380.
Fixes #57349.

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review

**Review by cston — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_IndexerAccess.cs [hunk 1]
```diff
@@ -312,7 +398,6 @@ private BoundSequence VisitIndexOrRangePatternIndexerAccess(BoundIndexOrRangePat
             var sideEffectsBuilder = ArrayBuilder<BoundExpression>.GetInstance();
 
             var receiverLocal = F.StoreToTemp(VisitExpression(receiver), out var receiverStore);
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_IndexerAccess.cs:400]: It looks like the comment on 387 is no longer accurate.

**Review by RikkiGibson — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_IndexerAccess.cs (hunk 1 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_IndexerAccess.cs:400]: > It looks like the comment on 387 is no longer accurate.

Are you suggesting swapping the following lines?
```
            // int length = receiver.length;
            // Range range = argumentExpr;
```

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_IndexerAccess.cs:400]: Yes, it looks like that's all that would need to change to make the comment accurate.

**Comment by hez2010** [comment]
Will this ship with .NET 6 service upgrade?

**Comment by jcouv** [comment]
@hez2010 I think the next .NET SDK version to pick this change will be 6.1 (not a servicing release of 6.0).

**Comment by jcouv** [comment]
@hez2010 I got the numbering scheme wrong. The next non-servicing version for .NET SDK will be 6.0.200 (we just shipped 6.0.100 and servicing will stay in the 100 range).

**Pull request merged**